### ISSUE TITLE
Support wildcard TLS certificates

### ISSUE THREAD

**Issue by OCTAGRAM** [AdaCore/aws#177]
Let's Encrypt provides free wildcard certificates, but each certificate for single common name only. No way to bundle 50 domains into 1 default certificate. Have to issue 100 certificates, one for each domain and one wildcard for all subdomains, plus 3- and 4-level wildcard where desired. AWS needs to pick appropriate one via SNI, but it fails currently.

This is my patch proposal.

````diff -Nur /root/GNAT2017/aws.no_wildcard/config/ssl/aws-net-ssl__openssl.adb /root/GNAT2017/aws/config/ssl/aws-net-ssl__openssl.adb
--- /root/GNAT2017/aws.no_wildcard/config/ssl/aws-net-ssl__openssl.adb  2020-09-04 18:14:35.086334886 +0200
+++ /root/GNAT2017/aws/config/ssl/aws-net-ssl__openssl.adb      2020-09-04 18:27:39.000000000 +0200
@@ -39,6 +39,7 @@
 with Ada.Command_Line;
 with Ada.Containers.Indefinite_Hashed_Maps;
 with Ada.Directories;
+with Ada.Strings.Fixed;
 with Ada.Strings.Equal_Case_Insensitive;
 with Ada.Strings.Hash_Case_Insensitive;
 with Ada.Task_Attributes;
@@ -1088,6 +1089,7 @@
       pragma Unreferenced (ad);
       use C.Strings;
       Server_Name : constant chars_ptr := TSSL.SSL_get_servername (Session);
+      Server_Name_String : constant String := Value (Server_Name);
       CH : Host_Certificates.Cursor;
       Dummy : TSSL.SSL_CTX;
    begin
@@ -1095,7 +1097,27 @@
          return TSSL.SSL_TLSEXT_ERR_OK;
       end if;

-      CH := Hosts.Find (Value (Server_Name));
+      CH := Hosts.Find (Server_Name_String);
+
+      if not Host_Certificates.Has_Element (CH) then
+         declare
+            First_Dot_Position : constant Natural :=
+              Ada.Strings.Fixed.Index (Server_Name_String, ".");
+         begin
+            if First_Dot_Position > 0 then
+               declare
+                  Wildcard_Pattern : String
+                    (1 .. Server_Name_String'Last + 2 - First_Dot_Position);
+               begin
+                  Wildcard_Pattern (1) := '*';
+                  Wildcard_Pattern (2 .. Wildcard_Pattern'Last) :=
+                    Server_Name_String
+                      (First_Dot_Position .. Server_Name_String'Last);
+                  CH := Hosts.Find (Wildcard_Pattern);
+               end;
+            end if;
+         end;
+      end if;

       if Host_Certificates.Has_Element (CH) then
          Dummy := TSSL.SSL_set_SSL_CTX (Session, Hosts.all (CH));````

**Comment by TurboGit** [comment]
This makes sense ! Please can you provide a PR based on this patch ? TIA.

**Comment by anisimkov** [comment]
Ivan, could you pleas close this issue. The feature is implemented in 71417339e43ba61ce25e7dd0fa39c0f5b656ab42
